cpp/include/raft/core/serialization.hpp

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <raft/core/handle.hpp>
+#include <raft/detail/serialization.hpp>
+
+#include <vector>
+
+namespace raft {
+
+/**
+ * @brief Write a serializable state of an object to memory.
+ *
+ * @tparam T type of the serializable object
+ * @tparam ContextArgs types of context required for serialization
+ *
+ * @param[out] out a host pointer to the location where to store the state;
+ *   when `nullptr`, the actual data is not written (only the size-to-written is calculated).
+ * @param obj the object to be serialized
+ * @param args context required for serialization
+ * @return the number of bytes (to be) written by the pointer
+ */
+template <typename T, typename... ContextArgs>
+auto serialize(uint8_t* out, const T& obj, ContextArgs&&... args) -> size_t
+{
+  return detail::call_serialize<T, ContextArgs...>(out, obj, std::forward<ContextArgs>(args)...);
+}
+
+/**
+ * @brief Write a serializable state of an object to a host vector.
+ *
+ * @tparam T type of the serializable object
+ * @tparam ContextArgs types of context required for serialization
+ *
+ * @param obj the object to be serialized
+ * @param args context required for serialization
+ * @return the serialized state
+ */
+template <typename T, typename... ContextArgs>
+auto serialize(const T& obj, ContextArgs&&... args) -> std::vector<uint8_t>
```

Comment:
The base implementation (`to_bytes`) at the moment writes to existing storage, but I don't have any checks whatsoever to see if there is enough space allocated. In the current state, the user is supposed to call `serialize` with `nullptr` to find out how much space it needs.

Reply:
@achirkin Would you consider a design that names the space query as a separate function from serialization?  Compare to cuSPARSE, which separates `*_bufferSize` functions from `*_solve` functions.

Reply:
I'd be totally fine separating them, though for some reason I've had an impression that in raft/cuml we follow the CUB approach of using the same function. @cjnolet , which one do you prefer?

Reply:
This one is hard. @mhoemmen, we've been trying to keep the APIs in RAFT a little more C++ friendly, so generally trying to avoid the two-stage `compute buffer size` -> `perform computation` where at all possible. I'm usually not particularly in love w/ returning objects from functions either, unless we are explicitly using something like a factory for a complex object, because then the user cannot control the memory allocation. 

In this particular case, if we are only ever expecting the serialized buffer to be in host memory, I could see an argument for having serialize return a vector, or even accepting a vector and populating it. I think if I was to choose between the various options presented, I most like the idea of returning/populating some sort of streamable/iterable type if we can. I find that to be the most flexible design because a large stream can go right to disk and have to buffer only a small portion into memory at any given time. Further, the stream is also nice because you can collect it into a byte buffer at any point if needed.

Reply:
@cjnolet wrote:
> I think if I was to choose between the various options presented, I most like the idea of returning/populating some sort of streamable/iterable type if we can.

I do like the idea of writing to some kind of output range.  If it's a random access range, it would be nice to test if it's long enough and report an error if not, rather than just clobbering it.  This is always detectable, and is generally recoverable (allocate more memory) or at least worth reporting ("insufficient disk space").

Please do consider not allocating the return buffer each time.  This would make it impossible to write sequential records to contiguous storage.  It can also be surprisingly expensive in a tight loop.

Reply:
> Please do consider not allocating the return buffer each time. This would make it impossible to write sequential records to contiguous storage. It can also be surprisingly expensive in a tight loop.

I do wholeheartedly agree with this.